File: src/main/java/singleton_test/LazyNotThreadSafe.java
package singleton_test;

/**
 * Created by Alex on 2017/5/27.
 * 2.懒加载，但是线程不安全的单例
 * Alex
 */

public class LazyNotThreadSafe {
    private static LazyNotThreadSafe lazyNotThreadSafe;

    private LazyNotThreadSafe() {
        System.out.println("懒加载，但是线程不安全的单例");
    }

    public static LazyNotThreadSafe instances() {
        if (lazyNotThreadSafe == null) lazyNotThreadSafe = new LazyNotThreadSafe();
        return lazyNotThreadSafe;
    }
}

File: src/main/kotlin/kotlin_singleton_test/LazyNotThreadSafe.kt
package kotlin_singleton_test
/**
 * Created by Alex on 2017/5/27.
 * 2.懒加载，但是线程不安全的单例
 * Alex
 */

public class LazyNotThreadSafe private constructor(msg:String){

    init {
        println(msg)
    }

    companion object {
        //kotlin中的常规写法
        val instances by lazy(LazyThreadSafetyMode.NONE) {
            LazyNotThreadSafe("kotlin：懒加载，但是线程不安全的单例")
        }
        //模仿java中的写法
        private var instance2: LazyNotThreadSafe? = null;

        fun instances2(): LazyNotThreadSafe {
            if (instance2 == null) {
                instance2 = LazyNotThreadSafe("kotlin模仿JAVA：懒加载，但是线程不安全的单例")
            }
            return instance2!!
        }
    }


}